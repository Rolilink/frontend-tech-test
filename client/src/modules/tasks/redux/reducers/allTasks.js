import { handleActions } from 'redux-actions';
import successCase from './cases/allTasks/fetchTasksSuccess';
import {
  setLoading as loadingCase,
  setError as errorCase,
} from './cases/http';
import { state as stateFactory } from '../factories';

import {
  FETCH_TASKS_LOADING,
  FETCH_TASKS_SUCCESS,
  FETCH_TASKS_ERROR,
} from '../actionTypes';

const reducer = handleActions({
  [FETCH_TASKS_SUCCESS]: successCase,
  [FETCH_TASKS_LOADING]: loadingCase,
  [FETCH_TASKS_ERROR]: errorCase,
}, stateFactory());

export default reducer;
