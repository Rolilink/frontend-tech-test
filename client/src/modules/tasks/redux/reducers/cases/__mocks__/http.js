export const setSuccess = jest.fn(() => ({}));
export const setLoading = jest.fn(() => ({}));
export const setError = jest.fn(() => ({}));
