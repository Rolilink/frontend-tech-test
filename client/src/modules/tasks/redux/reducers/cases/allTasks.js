import { allTasks } from '../../factories';
import normalizeTasks from '../../utils';

export default (state, { payload }) => ({
  ...state,
  ...allTasks({ tasks: { ...state.tasks, ...normalizeTasks(payload.tasks) } }),
});
