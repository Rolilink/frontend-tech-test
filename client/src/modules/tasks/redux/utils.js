export default (tasks) => {
  const normalizedTasks = {};

  tasks.forEach((task) => {
    normalizedTasks[task.id] = task;
  });

  return normalizedTasks;
};
