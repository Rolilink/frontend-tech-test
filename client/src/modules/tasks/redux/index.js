import * as actionTypes from './actionTypes';
import * as factories from './factories';
import * as selectors from './selectors';
import * as utils from './utils';

import { fetchTasks, createTask } from './actions';

import reducerModule from './reducers';

const reducer = {
  name: 'tasks', reducer: reducerModule,
};

export {
  fetchTasks,
  createTask,
  actionTypes,
  factories,
  selectors,
  reducer,
  utils,
};
