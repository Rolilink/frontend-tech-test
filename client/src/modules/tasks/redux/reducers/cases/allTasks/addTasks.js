/*
 This case adds one or more tasks through an array into the tasks state of all
 tasks, it is created this way to be reused through fetch, update and create.
*/
import { allTasks } from '../../../factories/';
import { normalizeTasks } from '../../../utils';


const tasks = allTasks.tasks;

export default (state, { payload }) => ({
  ...state,
  ...tasks({ tasks: { ...state.tasks, ...normalizeTasks(payload.tasks) } }),
});
