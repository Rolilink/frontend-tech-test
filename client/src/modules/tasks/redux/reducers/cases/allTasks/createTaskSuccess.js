import addTasks from './addTasks';
import { setSuccess } from '../http';

export default(state, action) => ({
  ...addTasks(
    state,
    { ...action, payload: { tasks: [action.payload.task] } }),
  isFetching: setSuccess(state, action).isFetching,
  error: setSuccess(state, action).error,
});
