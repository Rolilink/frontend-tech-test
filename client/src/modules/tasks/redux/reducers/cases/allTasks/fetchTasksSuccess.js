/*
 This case is called when multiple task are fetched 
*/

import addTasks from './addTasks';
import { setSuccess } from '../http';

export default(state, action) => ({
  ...addTasks(state, action),
  isFetching: setSuccess(state).isFetching,
  error: setSuccess(state).error,
});
