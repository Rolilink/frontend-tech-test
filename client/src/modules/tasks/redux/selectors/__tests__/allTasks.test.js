import { taskFactory } from '../../../../../__tests__/testUtils/tasks';
import { httpFactory } from '../../../../../__tests__/testUtils/http';
import { normalizeTasks } from '../../utils';
import { getTasks, getIsFetching, getError } from '../allTasks';

const state = {
  tasks: {
    allTasks: {
      tasks: {
        ...normalizeTasks([
          taskFactory({ id: 1 }),
          taskFactory({ id: 2 }),
          taskFactory({ id: 3 }),
        ]),
      },
      ...httpFactory(
        {
          error: 'error',
          isFetching: true,
        },
      ),
    },
  },
};

describe('Tasks.Redux.Selectors.AllTasks.getTasks', () => {
  it('should get all tasks when receiving a state', () => {
    expect(getTasks(state)).toMatchSnapshot();
  });
});

describe('Tasks.Redux.Selectors.AllTasks.getIsFetching', () => {
  it('should get isFetching from AllTasks when receiving a state', () => {
    expect(getIsFetching(state)).toBe(true);
  });
});

describe('Tasks.Redux.Selectors.AllTasks.getError', () => {
  it('should get the error from allTasks when receiving a state', () => {
    expect(getError(state)).toBe('error');
  });
});
