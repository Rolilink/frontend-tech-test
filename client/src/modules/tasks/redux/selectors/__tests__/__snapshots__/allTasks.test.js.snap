// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks.Redux.Selectors.AllTasks.getTasks should get all tasks when receiving a state 1`] = `
Object {
  "1": Object {
    "description": "Description",
    "id": 1,
    "title": "Title",
  },
  "2": Object {
    "description": "Description",
    "id": 2,
    "title": "Title",
  },
  "3": Object {
    "description": "Description",
    "id": 3,
    "title": "Title",
  },
}
`;
